src/Util/PhpServicesCreator.php

```diff
@@ -0,0 +1,739 @@
+
+        $closureStmt = $this->createClosureStmts((new Parser())->parse($yaml, Yaml::PARSE_CUSTOM_TAGS), $creatorFactory);
+
+        $creatorFactory->addStmt(
+            new Node\Stmt\Return_(
+                new Node\Expr\Closure([
+                    'params' => [
+                        (new Param('configurator'))->setType('ContainerConfigurator')->getNode(),
+                    ],
+                    'stmts' => $closureStmt,
+                ])
+            )
+        );
+
+        /** @var Node\Stmt\Namespace_ $node */
+        $node = $creatorFactory->getNode();
+        $beforeClosure = !empty($this->useStatements) ? "\n" : null;
+        // add a blank line between the last use statement and the closure
+        array_splice($node->stmts,
+            -1,
+            0,
+            [new Node\Name($beforeClosure."/*\n * This file is the entry point to configure your own services.\n */")]
+        );
+
+        return (new Standard())->prettyPrintFile([$node])."\n";
+    }
+
+    /**
+     * @return Node[]
+     */
+    private function createClosureStmts(array $yamlData, $creatorFactory): array
+    {
+        foreach ($yamlData as $key => $values) {
+            if (null === $values) {
+                // declare the variable ($parameters/$services) even if the key is written without values.
+                $values = [];
+            }
+            switch ($key) {
+                case 'parameters':
+                    $this->addParametersNodes($values);
+                    break;
+                case 'imports':
+                    $this->addImportsNodes($values);
+                    break;
+                case 'services':
+                    $this->addServicesNodes($values);
+                    break;
+                default:
+                    throw new \LogicException(sprintf('The key %s is not supported by the converter: only service config can live in services.yaml for conversion.', $key));
+                    break;
+            }
+        }
+
+        sort($this->useStatements);
+        foreach ($this->useStatements as $className) {
+            $creatorFactory->addStmt($this->factory->use($className));
+        }
+
+        // remove the last carriage return "\n" if exists.
+        $lastStmt = $this->stmts[array_key_last($this->stmts)];
+        $lastStmt->parts[0] = rtrim($lastStmt->parts[0], "\n");
+
+        return $this->stmts;
+    }
+
+    private function addParametersNodes(array $parameters)
+    {
+        $this->addLineStmt('// Put parameters here that don\'t need to change on each machine where the app is deployed.');
+        $this->addLineStmt('// https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration.');
+
+        $this->addLineStmt('$parameters = $configurator->parameters();', true);
+
+        foreach ($parameters as $parameterName => $value) {
+            $this->addLineStmt(sprintf(
```

Comment:
Actually, all these `sprintf(%variable, %call, %args)` should not be string, but a node.

Using string in a node as `Name`, while it is  `MethodCall` or `Assing` is like writing PHP in one big string.

Reply:
That's 100% true. I think, at some point (and I think it was quite pragmatic) @archeoprog took some shortcuts, where writing string code was simpler than using the Nodes. In this case (causeI just tried it), it's easy until the argument part, which has a lot of complex logic attached to it - specifically the `$this->toString()` method. 

So I don't think it's worth the work to refactor the internals of the class, though you do have a perfectly valid point.

Reply:
It makes sense as 1st draft with lack of knowledge of php-parser, I'm :+1: for that.

Just now it would deserve using nodes correctly. As this sprintf magic disables php-parser printer features. It's too much to think about on every line, instead of extending php-parser printer on 1 place :) (I'm super lazy btw)

E.g. this is how you add indentation to all fluent method calls:

https://github.com/migrify/migrify/pull/45/files#diff-3079c82962d6f72f03af70c687388891R17

Reply:
> Actually, all these `sprintf(%variable, %call, %args)` should not be string, but a node.
> 
> Using string in a node as `Name`, while it is `MethodCall` or `Assing` is like writing PHP in one big string.

That's exactly what I wanted to do with Name, it's a little "hack" reserved for the content of the Closure because the use of nodes proved to be, at this level, a waste of time, very tedious where the manipulation of strings was much simpler.  But this is only my personal vision of complexity and now that you have a few sorting algorithms as a basis, feel free to adapt for your project and see what happens with the nodes.  I'm sure @weaverryan would be interested to see the result :)

Reply:
Yeah, it was my first project with the bookstore.  I tried a refactoring with knots.  Even without knowing well, the Dump system allows to get nodes corresponding to php instructions so I had this help :)



In spite of that, I couldn't find the interest to continue, surely it would have been necessary to start again from the beginning, impossible for me.

Reply:
Got you. It makes sense :+1: 

I look forward to your final version.